src/transaction/log_recovery_redo_parallel.hpp

```diff
@@ -408,16 +411,63 @@ namespace cublog
       // by design pointer is guaranteed to outlive this instance
       const log_lsa *const m_end_redo_lsa;
       const log_reader::fetch_mode m_log_reader_page_fetch_mode;
+      reusable_jobs_stack *const m_reusable_job_stack;
 
       LOG_RECTYPE m_log_rtype;
   };
 
 
+  /* a preallocated pool of jobs
+   */
+  class reusable_jobs_stack final
+  {
+    public:
+      reusable_jobs_stack ();
+      reusable_jobs_stack (const reusable_jobs_stack &) = delete;
+      reusable_jobs_stack (reusable_jobs_stack &&) = delete;
+
+      ~reusable_jobs_stack ();
+
+      reusable_jobs_stack &operator = (const reusable_jobs_stack &) = delete;
+      reusable_jobs_stack &operator = (reusable_jobs_stack &&) = delete;
+
+      void initialize (std::size_t a_stack_size, const log_lsa *a_end_redo_lsa,
+		       bool force_each_page_fetch);
+      /* used for unit testing */
+      void initialize (std::size_t a_stack_size,
+		       std::function<redo_parallel::redo_job_base * ()> a_job_factory);
+
+      std::size_t size () const
+      {
+	return m_stack_size;
+      }
+
+      redo_parallel::redo_job_base *blocking_pop ();
+      void push (redo_parallel::redo_job_base *a_job);
+
+    private:
+      std::size_t m_stack_size;
+
+      /* pop is done, unsynchronized, from this stack
+       * if empty, it is re-filled, synchronized, from the push stack
+       */
+      std::stack<redo_parallel::redo_job_base *> m_pop_stack;
+      /* push happens, synchronized, on this stack
+       */
+      std::stack<redo_parallel::redo_job_base *> m_push_stack;
+      std::mutex m_mutex;
+      /* used to wait for jobs to be available on the push stack
+       * in case no jobs are found on the pop stack
+       */
+      std::condition_variable_any m_jobs_available_on_push_stack_cv;
```

Comment:
Quote from [std::condition_variable_any](https://en.cppreference.com/w/cpp/thread/condition_variable_any):

> If the lock is std::unique_lock\<std::mutex\>, std::condition_variable may provide better performance.

```suggestion
      std::condition_variable m_jobs_available_on_push_stack_cv;
```

Reply:
Autocompletion mistake I guess. Fixed.